plugins/FrequencyShifter/FrequencyShifterEffect.cpp

```diff
@@ -0,0 +1,294 @@
+	Effect(&frequencyshifter_plugin_descriptor, parent, key),
+	m_controls(this)
+{
+	m_ringBuf.resize(1, {0.f, 0.f});
+
+	connect(Engine::audioEngine(), &AudioEngine::sampleRateChanged,
+		this, &FrequencyShifterEffect::updateSampleRate);
+	updateSampleRate();
+}
+
+Effect::ProcessStatus FrequencyShifterEffect::processImpl(SampleFrame* buf, const fpp_t frames)
+{
+	const float twoPi = std::numbers::pi_v<float> * 2.0f;
+
+	const float mix = m_controls.m_mix.value() * wetLevel();
+	const float fs = m_controls.m_freqShift.value();
+	const float spread = m_controls.m_spreadShift.value();
+	const float ring = m_controls.m_ring.value();
+	const float feedback = m_controls.m_feedback.value();
+	const float delayLen = (m_controls.m_delayLengthLong.value() + m_controls.m_delayLengthShort.value()) * 0.001f * m_sampleRate;
+	const float delayDamp = m_controls.m_delayDamp.value();
+	const float delayGlide = m_controls.m_delayGlide.value();
+	const float lfoAmt = m_controls.m_lfoAmount.value();
+	const float lfoRate = (m_controls.m_lfoRate.value() / m_sampleRate) * twoPi;
+	const float lfoSt = m_controls.m_lfoStereoPhase.value() * twoPi;
+	const bool antireflect = m_controls.m_antireflect.value();
+	const int routeMode = m_controls.m_routeMode.value();
+	const float harmonics = m_controls.m_harmonics.value();
+	const float glide = m_controls.m_glide.value();
+	const float tone = m_controls.m_tone.value();
+	const float phase = m_controls.m_phase.value() * twoPi;
+
+	const bool resetShifterBtn = m_controls.m_resetShifter.value();
+	const bool resetLfoBtn = m_controls.m_resetLfo.value();
+
+	if (!m_prevResetShifter && resetShifterBtn)
+	{
+		m_phase[0] = 0.f;
+		m_phase[1] = 0.f;
+	}
+	if (!m_prevResetLfo && resetLfoBtn)
+	{
+		m_lfoPhase = 0.f;
+	}
+	m_prevResetShifter = resetShifterBtn;
+	m_prevResetLfo = resetLfoBtn;
+
+	const float twoPiOverSr = twoPi / m_sampleRate;
+	const float invRing = 1.f - ring;
+	const bool parallelFB = (routeMode >= 1);
+	const bool routeAdd = (routeMode == 1);
+
+	const bool doHarm = (harmonics > 0.f);
+	const float harmFactor = harmonics * 20.f + 1.f;
+	const float harmDiv = 1.f / (harmonics * 0.3f + 1.f);
+
+	const float dampCoeff = std::exp(-2.f * std::numbers::pi_v<float> * delayDamp / m_sampleRate);
+	const float toneCoeff = std::exp(-2.f * std::numbers::pi_v<float> * tone / m_sampleRate);
+	const float delayGlideCoeff = delayGlide ? std::exp(-1.f / (delayGlide * m_sampleRate)) : 0.f;
+	const float glideCoeff = glide ? std::exp(-1.f / (glide * m_sampleRate)) : 0.f;
+
+	const int wantSize = static_cast<int>(delayLen) + 4;// +4 provides space for interpolation
+	if (wantSize > m_ringBufSize)
+	{
+		m_ringBufSize = wantSize;
+		m_ringBuf.resize(wantSize);
+	}
+
+	// we only bother with wrapping phases once per buffer
+	m_lfoPhase = std::fmod(m_lfoPhase, twoPi);
+	for (int ch = 0; ch < 2; ++ch)
+	{
+		m_phase[ch] = std::fmod(m_phase[ch], twoPi);
+	}
+
+	for (fpp_t i = 0; i < frames; ++i)
```

Comment:
`i` is a frame count, not a measure of frames-per-period, no?

```suggestion
	for (f_cnt_t i = 0; i < frames; ++i)
```

Both are just `size_t` at the end of the day, which IMO is also acceptable.